modal/image.py

```diff
@@ -211,10 +250,12 @@ def _deps() -> List[_Object]:
             return deps
 
         async def _load(self: _Image, resolver: Resolver, existing_object_id: Optional[str]):
+            builder_version = await _get_builder_version()
```

Comment:
I think this should just make a single RPC per Python process? If you look in `_get_builder_version` we do cache the version.

Reply:
That said — good point about `Resolver.client`; in the original design we were doing this elsewhere. Though is it sort of weird to store this metadata on the `Client`? I'm not sure how to weight the tradeoff between having an extra RPC and accessing / storing this information in an obvious place.

Reply:
Yeah fair enough! I guess, data storage in the client library is temporary, but RPC protocol updates are forever

Reply:
If we wish to avoid this entirely, does it make sense to return the image version for the workspace in `ClientHello` and store it on the client, or in `AppCreate` and store it on the `stub`?

Reply:
If we set it on the client during `ClientHello` then it would be available via `resolve._client` as Eric noted below. (Private attribute though — implementation detail?). It's not obvious to me whether we have access to the stub in this scope?

The other thing to consider, I guess, is whether it's a good idea to add additional server-side operations during _every_ `ClientHello`. I confess that I don't have a great intuition for the relative magnitudes we're talking about here.

From a pure code organization point of view, I would say that this information seems a bit out of place in both `ClientHelloResponse` and `AppCreateResponse`, given the current definition of those protos. Although this implementation is slightly clunky too, given the way we're handling caching.

Reply:
> via `resolve._client` as Eric noted below. (Private attribute though

Just wanted to mention that it's a read-only public property, `resolver.client`. It's here because the resolver is intended to be the object that passes around a client for load() methods to use!

Reply:
Thanks, didn’t read far enough down

Reply:
OK after further discussion, I am thinking that saving one round trip during local operations isn't worth adding an unexpected dependency between image creation and client credential validation, especially as there are a few routes that we can take to create a client _without_ going through `ClientHello`.